File: src/voyage_geo/stages/analysis/analyzers/citation.py
"""Citation analyzer — detects URLs and sources cited in responses."""

from __future__ import annotations

import re
from collections import Counter
from urllib.parse import urlparse

from voyage_geo.types.analysis import CitationScore, CitationSource
from voyage_geo.types.brand import BrandProfile
from voyage_geo.types.result import QueryResult

URL_PATTERN = re.compile(r"https?://[^\s<>\"')\]]+")


class CitationAnalyzer:
    name = "citation"

    def analyze(self, results: list[QueryResult], profile: BrandProfile) -> CitationScore:
        valid = [r for r in results if not r.error and r.response]
        if not valid:
            return CitationScore()

        source_counter: Counter[str] = Counter()
        source_providers: dict[str, set[str]] = {}
        responses_with_citations = 0

        for r in valid:
            urls = URL_PATTERN.findall(r.response)
            if urls:
                responses_with_citations += 1
            for url in urls:
                try:
                    domain = urlparse(url).netloc
                    if domain:
                        source_counter[domain] += 1
                        source_providers.setdefault(domain, set()).add(r.provider)
                except Exception:
                    pass

        total = sum(source_counter.values())
        citation_rate = responses_with_citations / len(valid) * 100 if valid else 0

        by_provider: dict[str, int] = {}
        provider_groups: dict[str, list[QueryResult]] = {}
        for r in valid:
            provider_groups.setdefault(r.provider, []).append(r)
        for prov, prov_results in provider_groups.items():
            by_provider[prov] = sum(1 for r in prov_results if URL_PATTERN.search(r.response))

        top_sources = [
            CitationSource(
                source=domain,
                count=count,
                providers=sorted(source_providers.get(domain, set())),
            )
            for domain, count in source_counter.most_common(10)
        ]

        return CitationScore(
            total_citations=total,
            unique_sources_cited=len(source_counter),
            citation_rate=round(citation_rate, 1),
            by_provider=by_provider,
            top_sources=top_sources,
        )

File: src/stages/analysis/analyzers/citation.ts
import type { Analyzer } from './types.js';
import type { QueryResult } from '../../../types/result.js';
import type { BrandProfile } from '../../../types/brand.js';
import type { CitationScore, CitationSource } from '../../../types/analysis.js';
import { groupBy, percentage } from '../statistics.js';

const URL_REGEX = /https?:\/\/[^\s)<>]+/g;

export class CitationAnalyzer implements Analyzer {
  name = 'citation';

  analyze(results: QueryResult[], _profile: BrandProfile): CitationScore {
    const validResults = results.filter((r) => !r.error && r.response);

    const sourceCounts = new Map<string, { count: number; providers: Set<string> }>();
    let totalCitations = 0;
    let responsesWithCitations = 0;

    for (const result of validResults) {
      const urls = result.response.match(URL_REGEX) || [];
      if (urls.length > 0) {
        responsesWithCitations++;
      }

      for (const url of urls) {
        totalCitations++;
        try {
          const hostname = new URL(url).hostname.replace(/^www\./, '');
          const existing = sourceCounts.get(hostname) || { count: 0, providers: new Set() };
          existing.count++;
          existing.providers.add(result.provider);
          sourceCounts.set(hostname, existing);
        } catch {
          // Invalid URL, skip
        }
      }
    }

    const citationRate = percentage(responsesWithCitations, validResults.length);

    // By provider
    const byProvider: Record<string, number> = {};
    const providerGroups = groupBy(validResults, (r) => r.provider);
    for (const [provider, providerResults] of providerGroups) {
      let providerCitations = 0;
      for (const result of providerResults) {
        const urls = result.response.match(URL_REGEX) || [];
        providerCitations += urls.length;
      }
      byProvider[provider] = providerCitations;
    }

    // Top sources
    const topSources: CitationSource[] = Array.from(sourceCounts.entries())
      .map(([source, data]) => ({
        source,
        count: data.count,
        providers: Array.from(data.providers),
      }))
      .sort((a, b) => b.count - a.count)
      .slice(0, 10);

    return {
      totalCitations,
      uniqueSourcesCited: sourceCounts.size,
      citationRate,
      byProvider,
      topSources,
    };
  }
}
